File: Temple.Domain/Entities/PR/Person.cs
using Craft.Domain;

namespace Temple.Domain.Entities.PR
{
    public class Person : IObjectWithGuidID, IObjectWithValidTime, IClonableObject
    {
        public Guid ArchiveID { get; set; }
        public DateTime Created { get; set; }
        public DateTime Superseded { get; set; }
        public DateTime Start { get; set; }
        public DateTime End { get; set; }

        public Guid ID { get; set; }

        public string FirstName { get; set; }

        public string? Surname { get; set; }

        public string? Nickname { get; set; }

        public string? Address { get; set; }

        public string? ZipCode { get; set; }

        public string? City { get; set; }

        public DateTime? Birthday { get; set; }

        public string? Category { get; set; }

        public string? Description { get; set; }

        public bool? Dead { get; set; }

        public double? Latitude { get; set; }

        public double? Longitude { get; set; }

        public virtual ICollection<PersonComment>? Comments { get; set; }
        public virtual ICollection<PersonAssociation>? ObjectPeople { get; set; }
        public virtual ICollection<PersonAssociation>? SubjectPeople { get; set; }

        public IClonableObject Clone()
        {
            var clone = new Person();
            clone.CopyAttributes(this);
            return clone;
        }

        public override string ToString()
        {
            return $"{FirstName} ({Start} - {End})";
        }

        public void CopyAttributes(
            Person other)
        {
            ArchiveID = other.ArchiveID;
            Created = other.Created;
            Superseded = other.Superseded;
            Start = other.Start;
            End = other.End;
            ID = other.ID;
            FirstName = other.FirstName;
            Surname = other.Surname;
            Nickname = other.Nickname;
            Address = other.Address;
            ZipCode = other.ZipCode;
            City = other.City;
            Birthday = other.Birthday;
            Category = other.Category;
            Description = other.Description;
            Latitude = other.Latitude;
            Longitude = other.Longitude;
        }
    }
}

File: Temple.WebClient/src/app/models/person.ts
export interface Person {
  id: string;
  start: Date;
  end: Date;
  created: Date;
  superseded: Date;
  firstName: string;
  surname: string | null;
  nickname: string | null;
  address: string | null;
  zipCode: string | null;
  city: string | null;
  birthday: Date | null;
  category: string | null;
  description: string | null;
  dead: string | boolean | null;
  latitude: number | null;
  longitude: number | null;
}

export class Person implements Person {
  constructor(init?: PersonFormValues) {
    Object.assign(this, init); // Populate all of the properties that it can into our person
  }
}

export class PersonFormValues {
  id?: string = undefined;
  start: Date | null = null;
  end: Date | null = null;
  created: Date | null = null;
  superseded: Date | null = null;
  firstName: string = "";
  surname: string | null = "";
  nickname: string | null = "";
  address: string | null = "";
  zipCode: string | null = "";
  city: string | null = "";
  birthday: Date | null = null;
  category: string | null = "";
  description: string | null = "";
  dead: string | boolean | null = null;
  latitude: number | null = null;
  longitude: number | null = null;

  constructor(person?: PersonFormValues) {
    if (person) {
      this.id = person.id;
      this.start = person.start;
      this.end = person.end;
      this.created = person.start;
      this.superseded = person.end;
      this.firstName = person.firstName === null ? "" : person.firstName;
      this.surname = person.surname === null ? "" : person.surname;
      this.nickname = person.nickname === null ? "" : person.nickname;
      this.address = person.address === null ? "" : person.address;
      this.zipCode = person.zipCode === null ? "" : person.zipCode;
      this.city = person.city === null ? "" : person.city;
      this.birthday = person.birthday;
      this.category = person.category === null ? "" : person.category;
      this.description = person.description === null ? "" : person.description;
      this.dead = person.dead;
      this.latitude = person.latitude;
      this.longitude = person.longitude;
    }
  }
}
